compiler/src/dotty/tools/dotc/ast/Trees.scala

```diff
@@ -1366,13 +1366,19 @@ object Trees {
     /** The context to use when mapping or accumulating over a tree */
     def localCtx(tree: Tree)(using Context): Context
 
+    /** The context to use when transforming a tree.
+      * It ensures that the source information is correct.
+      * TODO: ensure transform is always called with the correct context as argument
+      * @see https://github.com/lampepfl/dotty/pull/13880#discussion_r836395977
+      */
+    def transformCtx(tree: Tree)(using Context): Context =
```

Comment:
`def transform` also does `inContext(localCtx(tree))` for various trees, could this logic be moved to `transformCtx` so InstrumentCoverage doesn't have to manually pass localCtx in various places? It should be possible to do it with less checks than in transform too (e.g. instead of checking for ValDef/DefDef/TypeDef, check for MemberDef)

Reply:
Sure!
Would that degrade the performance of `transform` (because we'll match twice on terms that need a localCtx) or not so much? (or we don't care?)

Reply:
Question about PackageDef:
I've changed
```scala
val pid1 = transformSub(tree.pid)
inContext(localCtx(tree)) {
  cpy.PackageDef(tree)(pid1, transformStats(stats, ctx.owner))
}
```
to
```scala
cpy.PackageDef(tree)(transformSub(pid), transformStats(stats, ctx.owner))
```
with the context set as `localCtx(tree)` by `transformCtx`.
It means that `pid` is now transformed inside of the local context. Is that a problem, given that it's only a RefTree to the package denotation?

Reply:
We can run the benchmarks on this PR to check that, but it's probably fine since we'll only check a few cases.

Reply:
>  It means that pid is now transformed inside of the local context. Is that a problem, given that it's only a RefTree to the package denotation?

I think that's fine, in fact it seems to match how mini-phases work: https://github.com/lampepfl/dotty/blob/15fccdeefdbe4aa0d10c57a3274da677edf0d7a8/compiler/src/dotty/tools/dotc/transform/MegaPhase.scala#L377-L382